### DISCUSSION TITLE
How to set fuses?

### DISCUSSION THREAD

**Discussion by A1essandro** [Alex079/vscode-avr-helper#20] [Q&A] [answered]
Hi all!

Could anyone please explain how to set fuses? I.e. I'm selecting lfuse after `Flash` and getting this error:

`The terminal process "bash '-c', '/usr/bin/avrdude -v -p attiny13 -c usbasp -Uflash:w:/home/a1essandro/Projects/avrtest/.vscode/avr.build/output.elf:e -Ulfuse:w:/home/a1essandro/Projects/avrtest/.vscode/avr.build/output.elf:e'" terminated with exit code: 1.`
File `output.elf` is created by `Build` so existing.

I want to change frequency of attiny from default to 9.6MHz, so as far as I understand value should be `lfuse:w:0x7A:m`.
During executing `Flash` and selecting `flash` I got this message:

```
avrdude: safemode: lfuse reads as 79
avrdude: safemode: hfuse reads as FF
avrdude: safemode: Fuses OK (E:FF, H:FF, L:79)

avrdude done.  Thank you.
```
So according to [calculator](http://eleccelerator.com/fusecalc/fusecalc.php?chip=attiny13) frequency is 4.8MHz.

Currently I'm setting fuses via CLI. It is not really convenuent.
Also my small question what's the point to specify device frequency?

**Comment by Alex079** [comment, marked as answer]
Hi!

You can use this small example for reference (attiny85 in this case):
```C++
#include <avr/io.h>

FUSES = {
  .low = 0xFF & FUSE_CKDIV8 & FUSE_SUT0 & FUSE_CKSEL0 & FUSE_CKSEL2 & FUSE_CKSEL3,
  .high = 0xFF & FUSE_SPIEN & FUSE_EESAVE,
  .extended = 0xFF
};

int main() { return 0; }
```
Once the FUSES are set in source code, corresponding "fuse" section will be present in ELF file after build. After this it should be possible to flash the fuses. I decided to do it this way in order to keep fuses configuration linked to source code.

The device frequency (F_CPU) is often needed for parts of code which depend on CPU frequency (i.e. cycle counter for delay). Some library code may rely on F_CPU to conditionally define parts of program. My idea is that leaving F_CPU optional may have adverse effect on compilation result, unlike fuses, which are optional.

I hope this helps.

**Reply by A1essandro** [reply]
Ok, thank you! I will try to set fuses via source code a bit later.
